### PR TITLE
Feature: plugin id filters & forgiving name filters

### LINKED ISSUE #24: Plugin filter for UI Template Tabs cannot filter by plugin id
Currently the plugin filter allows filtering by the exact display name of a plugin. However, this display name is intended for humans and may not be as stable as the plugin identifier.

- [ ] Add a filter type that can filter by the plugin id (without version)
- [ ] Update name filter to allow for minor spelling differences (i.e., convert everything to upper or lower case before comparing)
- [ ] Maybe make the name filter even more forgiving (e.g., match even if filter is only part of the name or use a fuzzy matching algorithm)

### LINKED ISSUE #27: Inconsistent plugin ids
When discovering plugins the qhana-plugin-registry assigns the `plugin_name` to the `plugin_id`:

https://github.com/UST-QuAntiL/qhana-plugin-registry/blob/a77126596bffd0118c44b99c0ba5c327f8bcf4d6/qhana_plugin_registry/db/util.py#L120

Shouldn't the [`RAMP.plugin_id`](https://github.com/UST-QuAntiL/qhana-plugin-registry/blob/a77126596bffd0118c44b99c0ba5c327f8bcf4d6/qhana_plugin_registry/db/models/plugins.py#L87) be the same as the plugin `identifier` from the qhana-plugin-runner, i.e. `<name>@<version>`?

Currently, the plugin filter implementation assumes that `RAMP.plugin_id` is of the form `<name>@<version>`. If a user specifies an id filter (#26) including `@<version>` but the `plugin_name` doesn't contain the version, the filter won't find the corresponding plugin.

### Regarding the Plugin Runner
While some plugins set the plugin name to the identifier (e.g. [SVM](https://github.com/UST-QuAntiL/qhana-plugin-runner/blob/3db676e5a950e55bffac32ee3f9a72bf24e6380d/stable_plugins/quantum_ml/qiskit_ml/svm/routes.py#L58)), most plugins use the name (e.g. [QKE](https://github.com/UST-QuAntiL/qhana-plugin-runner/blob/3db676e5a950e55bffac32ee3f9a72bf24e6380d/stable_plugins/quantum_ml/qiskit_ml/qiskit_quantum_kernel_estimation/routes.py#L56)). Which value is the intended one?

### PR DESCRIPTION
**Pull request by infacc**
- Adds the option to filter plugins by id
- Makes plugin name filters more forgiving using a fuzzy matching approach

Closes #24
Fixes #27 

Related to: https://github.com/UST-QuAntiL/qhana-ui/pull/62

### CODE REVIEW

File: qhana_plugin_registry/api/models/templates.py [hunk 1]
```diff
@@ -89,7 +89,7 @@ def validate_filter(filter_dict: dict, path: str = ""):
                         f"Invalid plugin filter: Invalid version '{v}'. Version must be a PEP 440 specifier (string). (Path: {current_path})"
                     )
                 try:
-                    SpecifierSet(v)
+                    Specifier(v)
```
Comment (buehlefs) [inline qhana_plugin_registry/api/models/templates.py:92]: Why not keep `SpecifierSet` here for added flexibility in specifying allowed versions? (Also: try to mention all relevant changes in the PR text, I would consider this one relevant for this PR)

File: qhana_plugin_registry/tasks/plugin_filter.py [hunk 2]
```diff
@@ -73,7 +76,24 @@ def get_plugins_from_filter(filter_dict: dict, plugin_mapping: dict) -> set[RAMP
...
+            # match plugin id or plugin id without version
+            plugin_ids = set()
+            for p_id, p in plugin_mapping.items():
+                if plugin_id == p.plugin_id:
+                    plugin_ids.add(p_id)
+                elif (
+                    "@" in p.plugin_id
+                    and plugin_id == p.plugin_id[: p.plugin_id.rfind("@")]
+                ):
+                    plugin_ids.add(p_id)
```
Comment (buehlefs) [inline qhana_plugin_registry/tasks/plugin_filter.py:95]: I would reverse the logic a bit:

```python
if '@' not in plugin_id and '@' in p.plugin_id:
    plugin_id += '@'  # <- ensures that the full id matches
    if p.plugin_id.startswith(plugin_id):
        # match
else:
    if p.plugin_id == plugin_id:
        # match
```

This should make this test easier to read.

Or split both ids at the '@' and compare from there?

**Review by buehlefs — CHANGES_REQUESTED** [review]
Maybe also add a filter criteria for the plugin type (i.e., processing, visualization, etc.) in this PR (or a seperate PR).
Are there other usefult filter criteria?

File: qhana_plugin_registry/api/models/templates.py (hunk 1 above)
Comment (infacc) [reply qhana_plugin_registry/api/models/templates.py:92]: Reasons for this change:
    - The plugin filter key 'version' suggests that only one version can be specified.
    - The `SpecifierSet` implementation could break the current test setup. Hypothesis might generate a test case with a version that's a valid `Specifier` but an invalid `SpecifierSet` (e.g. `'=====,0'`).

**I'll revert this change and adjust the unit test to work with `SpecifierSet`s.**

Comment (buehlefs) [reply qhana_plugin_registry/api/models/templates.py:92]: - in this context the "version" is used as a filter (where one usually wants to filter by version ranges and such)
- That are edge cases you will only find with such thorough tests. So 👍 for the test cases so far. If invalid specifier sets are rare, then you can filter those out after generating them with [assume](https://hypothesis.readthedocs.io/en/latest/details.html#making-assumptions)

File: qhana_plugin_registry/tasks/plugin_filter.py (hunk 2 above)
Comment (infacc) [reply qhana_plugin_registry/tasks/plugin_filter.py]: I implemented it using `split()`.

```python
if plugin_id == p.plugin_id:
    # match
elif plugin_id.split("@") == p.plugin_id.split("@")[:-1]:
    # match
```

For cases where ids have multiple `@`s and same prefixes, e.g. `asdf@123@v0`, `asdf@000@v1`, I matched the prefix until the last `@` for plugin id filters without version. Although we probably won't have such plugin ids in practice the hypothesis tests may generate them.

**Review by buehlefs — APPROVED** [review]
Just some small comments

File: tests/plugin_filter/test_plugin_filter.py [hunk 3]
```diff
@@ -227,9 +235,12 @@ def test_plugin_filter_name(tmp_db, client, template_tab, plugins, name: str):
     # test single name excluded filter
     filter_dict = {"not": {"name": name}}
     tab_plugin_ids = update_plugin_filter(tmp_db, client, template_tab, filter_dict)
-    filtered_plugin_ids = {
-        p_id for p_id, in tmp_db.session.query(RAMP.id).filter(RAMP.name != name).all()
-    }
+    filtered_plugin_ids = set()
+    for p_id, p_name in tmp_db.session.query(RAMP.id, RAMP.name).all():
+        sm = SequenceMatcher(None, p_name, name)
+        if sm.ratio() <= PLUGIN_NAME_MATCHING_THREASHOLD:
```
[the remark this answers is not in the document]
Comment (infacc) [reply tests/plugin_filter/test_plugin_filter.py:241]: According to the [documentation](https://docs.python.org/3/library/difflib.html#difflib.SequenceMatcher.ratio) `ratio()` is deterministic:
> Return a measure of the sequences’ similarity as a float in the range [0, 1].
Where T is the total number of elements in both sequences, and M is the number of matches, this is 2.0*M / T. Note that this is 1.0 if the sequences are identical, and 0.0 if they have nothing in common.

File: tests/plugin_filter/test_plugin_filter.py [hunk 4]
```diff
@@ -352,7 +367,7 @@ def test_plugin_filter_tag(tmp_db, client, template_tab, plugins, tag_name: str)
     ), f"filtering by multiple tags failed (filter: '{filter_dict}')"
 
 
-@settings(suppress_health_check=[HealthCheck.function_scoped_fixture], deadline=500)
+@settings(suppress_health_check=[HealthCheck.function_scoped_fixture], deadline=1000)
```
[the remark this answers is not in the document]
Comment (infacc) [reply tests/plugin_filter/test_plugin_filter.py:370]: I don't think so. The 2 relevant tests only rarely exceeded the time-limit (on my machine) and only with a debugger attached (conditional breakpoints set but didn't hit).

File: qhana_plugin_registry/tasks/plugin_filter.py [hunk 5]
```diff
@@ -89,14 +89,17 @@ def get_plugins_from_filter(
             for p_id, p in plugin_mapping.items():
                 if plugin_id == p.plugin_id:
                     plugin_ids.add(p_id)
-                elif plugin_id.split("@") == p.plugin_id.split("@")[:-1]:
+                elif (
+                    plugin_id.split("@") == p.plugin_id.split("@")[:-1]
+                ):  # id matches, except for the version string
```
Comment (buehlefs) [inline qhana_plugin_registry/tasks/plugin_filter.py:94]: If the autoformatter breaks a line like this it can be better to move the comment to the next line (into the else block)

**Review by buehlefs — COMMENTED** [review]
We need to adapt the plugin_id filter to the results from UST-QuAntiL/qhana-plugin-registry#27

**Review by buehlefs — APPROVED** [review]
